controller/rpc/scanner.go

```diff
@@ -3,193 +3,242 @@ package rpc
+	endpoint := mgr.getScannerEndpoint(scanner)
+	cluster.DeleteGRPCClient(endpoint)
+}
+
+func (mgr *ScannerAcquisitionManager) getScannerEndpoint(scanner *share.CLUSScanner) string {
 	return fmt.Sprintf("%s:%v", scanner.RPCServer, scanner.RPCServerPort)
 }
 
-func (mgr *ScanCreditManager) createScannerServiceWrapper(conn *grpc.ClientConn) cluster.Service {
+func (mgr *ScannerAcquisitionManager) createScannerServiceWrapper(conn *grpc.ClientConn) cluster.Service {
 	return share.NewScannerServiceClient(conn)
 }
 
-// shouldIncrementTask not apply for Ping, if the acquireScanCredit is called by ScanImage should be true
-func (mgr *ScanCreditManager) getScannerServiceClient(sid string, shouldIncrementTask bool) (share.ScannerServiceClient, string, error) {
-	mgr.mutex.Lock()
-	defer mgr.mutex.Unlock()
-
-	s, err := mgr.scannerLoadBalancer.GetScanner(sid)
+func (mgr *ScannerAcquisitionManager) getScannerServiceClient(sid string) (share.ScannerServiceClient, string, error) {
+	// Get scanner from KV store
+	scanner, _, err := mgr.clusterHelper.GetScannerRev(sid)
 	if err != nil {
 		log.WithFields(log.Fields{"error": err, "scanner": sid}).Error()
 		return nil, sid, err
 	}
 
 	// endpoint is also the key
-	endpoint := mgr.getScannerEndpoint(s.Scanner)
+	endpoint := mgr.getScannerEndpoint(scanner)
 	if cluster.GetGRPCClientEndpoint(endpoint) == "" {
 		if err := cluster.CreateGRPCClient(endpoint, endpoint, true, mgr.createScannerServiceWrapper); err != nil {
 			log.WithFields(log.Fields{"error": err}).Error("CreateGRPCClient")
 		}
 	}
 
 	c, err := cluster.GetGRPCClient(endpoint, nil, nil)
-	if err == nil {
-		if shouldIncrementTask {
-			s.AvailableScanCredits++
-		}
-		return c.(share.ScannerServiceClient), sid, nil
-	} else {
+	if err != nil {
 		log.WithFields(log.Fields{"error": err, "scanner": sid}).Error("Failed to connect to grpc server")
 		return nil, sid, err
 	}
+	return c.(share.ScannerServiceClient), sid, nil
 }
 
-func (mgr *ScanCreditManager) getAllAvailableScanners() map[string]share.CLUSScanner {
-	mgr.mutex.RLock()
-	defer mgr.mutex.RUnlock()
-
+func (mgr *ScannerAcquisitionManager) getAllAvailableScanners() map[string]share.CLUSScanner {
 	ret := map[string]share.CLUSScanner{}
 
-	for _, scanner := range mgr.scannerLoadBalancer.GetActiveScanners() {
-		if scanner != nil && scanner.Scanner != nil {
-			ret[scanner.Scanner.ID] = *scanner.Scanner
-		}
+	for _, scanner := range mgr.clusterHelper.GetAvailableScanners() {
+		ret[scanner.ID] = *scanner
 	}
 
 	return ret
 }
 
-func (mgr *ScanCreditManager) CountScanners() (busy, idle uint32) {
-	mgr.mutex.RLock()
-	defer mgr.mutex.RUnlock()
-
-	for _, s := range mgr.scannerLoadBalancer.GetActiveScanners() {
-		if s.AvailableScanCredits > 0 {
+func (mgr *ScannerAcquisitionManager) CountScanners() (busy, idle uint32) {
+	for _, scanner := range mgr.clusterHelper.GetAvailableScanners() {
+		if scanner.ScanCredit > 0 {
```

Comment:
scanner.ScanCredit represents the number of scan tasks that can still be assigned to this scanner.
Why `scanner.ScanCredit > 0` gets `busy++` & `scanner.ScanCredit <= 0` gets 'idle++'?